feat: add `blas/ext/sorthp`

headlessNode opened the pull request:
Resolves https://github.com/stdlib-js/metr-issue-tracker/issues/77.

## Description

> What is the purpose of this pull request?

This pull request:

-   add `blas/ext/sorthp`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/77
-   progresses #2656

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add blas/ext/sorthp

headlessNode commented on lib/node_modules/@stdlib/blas/ext/sorthp/lib/base.js:
```diff
@@ -0,0 +1,113 @@
+* limitations under the License.
+*/
+
+/* eslint-disable array-element-newline */
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/sorthp/lib/base.js:
```diff
@@ -0,0 +1,113 @@
+* var arr = ndarray2array( out );
+* // returns [ [ [ -5.0, -3.0 ] ], [ [ 1.0, 2.0 ] ], [ [ 4.0, 6.0 ] ] ]
+*/
+var sorthp = factory( table, [ idtypes0, idtypes1 ], odtypes, options ); // eslint-disable-line max-len
```
```suggestion
var sorthp = factory( table, [ idtypes0, idtypes1 ], odtypes, options );
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/sorthp/lib/base.js:
```diff
@@ -0,0 +1,112 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
```
```suggestion
```

headlessNode committed:
fix: lint error

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/sorthp/test/test.js:
```diff
@@ -0,0 +1,1390 @@
+var tape = require( 'tape' );
+var isndarrayLike = require( '@stdlib/assert/is-ndarray-like' );
+var isSameArray = require( '@stdlib/assert/is-same-array' );
+var isSameFloat64Array = require( '@stdlib/assert/is-same-float64array' );
+var Float64Array = require( '@stdlib/array/float64' );
```
```suggestion
```

kgryte committed:
chore: resolve lint failure

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/sorthp/docs/repl.txt:
```diff
@@ -0,0 +1,42 @@
+    Returns
+    -------
+    out: ndarray
+        Output array.
```
```suggestion
        Input array.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/sorthp/docs/repl.txt:
```diff
@@ -0,0 +1,42 @@
+
+{{alias}}( x[, sortOrder][, options] )
+    Sorts an input ndarray along one or more ndarray dimensions using heapsort.
```
```suggestion
    Sorts an input ndarray along one or more ndarray dimensions using heapsort.

    The function sorts an input ndarray in-place and thus mutates an input
    ndarray.
```

kgryte committed:
docs: add note

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: fix description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/sorthp/README.md:
```diff
@@ -0,0 +1,171 @@
+The function has the following parameters:
+
+-   **x**: input [ndarray][@stdlib/ndarray/ctor]. Must have a numeric or "generic" [data type][@stdlib/ndarray/dtypes].
+-   **sortOrder**: sort order (_optional_). May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor] having a real or "generic" [data type][@stdlib/ndarray/dtypes]. If provided an [ndarray][@stdlib/ndarray/ctor], the value must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the complement of the shape defined by `options.dims`. For example, give ...
```
I am wondering whether we should also support the string literals `'ascending'` and `'descending'` and then map those literals to `1` and `-1` internally before calling into lower level functions. The rationale is that, from a code readability standpoint, a number literal does not readily convey its meaning. Compare

```javascript
sorthp( x, -1 );
```

versus

```javascript
sorthp( x, 'descending' );
```

The latter is more literate.

This comment is only directed at the top-level end-user APIs. I am **not** suggesting we update any lower-level packages.

kgryte replied:
I am also okay supporting `'desc'` and `'asc'`, as well. Meaning, we could support four literal string values (but not ndarrays!): `descending`, `ascending`, `desc`, and `asc`.

kgryte replied:
Although, the abbreviated `desc` and `asc` are a bit less obvious.

kgryte replied:
@Planeshifter Do you have opinions here?

headlessNode replied:
@kgryte makes sense

headlessNode committed:
feat: add string support

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/ext/sorthp/README.md:
````diff
@@ -0,0 +1,195 @@
+// returns [ 2.0, -1.0, -3.0 ]
+```
+
+Alternatively, you can also provide a string literal to specify the sort order.
````
```suggestion
Alternatively, in addition to numeric values, you can specify the sort order via of the following string literals: `'ascending'`, `'asc'`, `'descending'`, or `'desc'`. The first two literals indicate to sort in ascending (i.e., increasing) order. The last two literals indicate to sort in descending (i.e., decreasing) order.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/sorthp/README.md:
````diff
@@ -0,0 +1,195 @@
+// returns [ 2.0, -1.0, -3.0 ]
+```
+
+Alternatively, in addition to numeric values, you can specify the sort order via of the following string literals: `'ascending'`, `'asc'`, `'descending'`, or `'desc'`. The first two literals indicate to sort in ascending (i.e., increasing) order. The last two literals indicate to sort in descending (i.e., decreasing) order.
````
```suggestion
In addition to numeric values, you can specify the sort order via of the following string literals: `'ascending'`, `'asc'`, `'descending'`, or `'desc'`. The first two literals indicate to sort in ascending (i.e., increasing) order. The last two literals indicate to sort in descending (i.e., decreasing) order.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/sorthp/README.md:
````diff
@@ -0,0 +1,195 @@
+// returns [ 2.0, -1.0, -3.0 ]
+```
+
+In addition to numeric values, you can specify the sort order via of the following string literals: `'ascending'`, `'asc'`, `'descending'`, or `'desc'`. The first two literals indicate to sort in ascending (i.e., increasing) order. The last two literals indicate to sort in descending (i.e., decreasing) order.
````
```suggestion
In addition to numeric values, you can specify the sort order via one of the following string literals: `'ascending'`, `'asc'`, `'descending'`, or `'desc'`. The first two literals indicate to sort in ascending (i.e., increasing) order. The last two literals indicate to sort in descending (i.e., decreasing) order.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/sorthp/README.md:
```diff
@@ -0,0 +1,195 @@
+
+-   **dims**: list of dimensions over which to perform operation. If not provided, the function performs the operation over all elements in a provided input [ndarray][@stdlib/ndarray/ctor].
+
+By default, the function uses an increasing sort order. To sort in a different order, provide a `sortOrder` argument.
```
```suggestion
By default, the function sorts elements in increasing order. To sort in a different order, provide a `sortOrder` argument.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/sorthp/README.md:
```diff
@@ -0,0 +1,195 @@
+
+-   The input [ndarray][@stdlib/ndarray/ctor] is sorted **in-place** (i.e., the input [ndarray][@stdlib/ndarray/ctor] is **mutated**).
+-   If `sortOrder < 0.0`, the input [ndarray][@stdlib/ndarray/ctor] is sorted in **decreasing** order. If `sortOrder > 0.0`, the input [ndarray][@stdlib/ndarray/ctor] is sorted in **increasing** order. If `sortOrder == 0.0`, the input [ndarray][@stdlib/ndarray/ctor] is left unchanged.
+-   If `sortOrder === asc` or `sortOrder === ascending`, the input [ndarray][@stdlib/ndarray/ctor] is sorted in **increasing** order. If `sortOrder === desc` or `sortOrder === descending`, the input [ndarray][@stdlib/ndarray/ctor] is sorted in **decreasing** order.
```
```suggestion
-   If `sortOrder < 0.0` or either `'desc'` or `'descending'`, the input [ndarray][@stdlib/ndarray/ctor] is sorted in **decreasing** order. If `sortOrder > 0.0` or either `'asc'` or `'ascending'`, the input [ndarray][@stdlib/ndarray/ctor] is sorted in **increasing** order. If `sortOrder == 0.0`, the input [ndarray][@stdlib/ndarray/ctor] is left unchanged.
```

kgryte committed:
docs: consolidate notes

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/sorthp/README.md:
````diff
@@ -0,0 +1,194 @@
+// returns [ 2.0, -1.0, -3.0 ]
+```
+
+In addition to numeric values, you can specify the sort order via one of the following string literals: `'ascending'`, `'asc'`, `'descending'`, or `'desc'`. The first two literals indicate to sort in ascending (i.e., increasing) order. The last two literals indicate to sort in descending (i.e., decreasing) order.
````
```suggestion
In addition to numeric values, one can specify the sort order via one of the following string literals: `'ascending'`, `'asc'`, `'descending'`, or `'desc'`. The first two literals indicate to sort in ascending (i.e., increasing) order. The last two literals indicate to sort in descending (i.e., decreasing) order.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
bench: avoid trivial sort

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM after clean-up.
